sdk/python/lib/pulumi/dynamic/__main__.py

```diff
@@ -34,9 +34,23 @@
 _GRPC_CHANNEL_OPTIONS = [("grpc.max_receive_message_length", _MAX_RPC_MESSAGE_SIZE)]
 
 
+_PROVIDER_CACHE: dict[str, ResourceProvider] = {}
+_PROVIDER_LOCK = Lock()
+
+
 def get_provider(props) -> ResourceProvider:
-    byts = base64.b64decode(props[PROVIDER_KEY])
-    return dill.loads(byts)
+    providerStr = props[PROVIDER_KEY]
+    provider = _PROVIDER_CACHE.get(providerStr)
+    if provider is None:
+        # This is pesimistic locking, because if two different providers try to fetch at the same time they
+        # serialise. But it means we don't create two instances of the same provider. Also looking at issues
+        # like https://github.com/pulumi/pulumi/issues/14159 there may be resource contention in dill.loads,
+        # that this locking strategy will reduce.
+        with _PROVIDER_LOCK:
+            byts = base64.b64decode(providerStr)
+            provider = dill.loads(byts)
+            _PROVIDER_CACHE[providerStr] = provider
```

Comment:
Perhaps I'm misunderstanding the comment, but should't we check the dict again while holding the lock?

```suggestion
            provider = _PROVIDER_CACHE.get(providerStr)
            if provider is None:
                byts = base64.b64decode(providerStr)
                provider = dill.loads(byts)
                _PROVIDER_CACHE[providerStr] = provider
```

Otherwise, if a bunch of callers are lined up waiting for the lock, they're each going to create a new instance.

Reply:
Ah yup, been a while since I had to write a manual check-lock-check method